src/repr/src/adt/macl_item.rs

```diff
@@ -0,0 +1,341 @@
+/// because we don't use OID as persistent identifiers for roles.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Hash, Deserialize)]
+pub struct MaclItem {
+    /// Role that this item grants privileges to.
+    pub grantee: RoleId,
+    /// Grantor of privileges.
+    pub grantor: RoleId,
+    /// Privileges bit flag.
+    pub acl_mode: AclMode,
+}
+
+impl MaclItem {
+    pub fn encode_binary(&self) -> Vec<u8> {
+        let mut res = Vec::new();
+        res.extend_from_slice(&self.grantee.encode_binary());
+        res.extend_from_slice(&self.grantor.encode_binary());
+        res.extend_from_slice(&self.acl_mode.bits().to_le_bytes());
+        res
+    }
+
+    pub fn decode_binary(raw: &[u8]) -> Result<MaclItem, Error> {
+        if raw.len() != MaclItem::binary_size() {
+            return Err(anyhow!(
+                "invalid binary size, expecting {}, found {}",
+                MaclItem::binary_size(),
+                raw.len()
+            ));
+        }
+
+        let role_id_size = RoleId::binary_size();
+
+        let grantee = RoleId::decode_binary(&raw[0..role_id_size])?;
+        let raw = &raw[role_id_size..];
+        let grantor = RoleId::decode_binary(&raw[0..role_id_size])?;
+        let raw = &raw[role_id_size..];
+        let acl_mode = u64::from_le_bytes(raw.try_into()?);
+
+        Ok(MaclItem {
+            grantee,
+            grantor,
+            acl_mode: AclMode { bits: acl_mode },
+        })
+    }
+
+    pub const fn binary_size() -> usize {
+        RoleId::binary_size() + RoleId::binary_size() + size_of::<AclMode>()
+    }
+}
+
+impl FromStr for MaclItem {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let re: Regex = Regex::new(&format!(
+            r"(.*)=(.+)/([{INSERT_CHAR},{SELECT_CHAR},
+        {UPDATE_CHAR},{DELETE_CHAR},{USAGE_CHAR},{CREATE_CHAR}]*)"
+        ))?;
+        let caps = re
+            .captures(s)
+            .ok_or_else(|| anyhow!("invalid maclitem '{s}'"))?;
+
+        let grantee: &str = caps.get(1).map_or("", |m| m.as_str());
+        let grantee = if grantee.is_empty() {
+            RoleId::Public
+        } else {
+            grantee.parse()?
+        };
+        let grantor: RoleId = caps
+            .get(2)
+            .ok_or_else(|| anyhow!("invalid maclitem '{s}'"))?
+            .as_str()
+            .parse()?;
+        let acl_mode: AclMode = caps
+            .get(3)
+            .ok_or_else(|| anyhow!("invalid maclitem '{s}'"))?
+            .as_str()
+            .parse()?;
+
+        Ok(MaclItem {
+            grantee,
+            grantor,
+            acl_mode,
+        })
+    }
+}
+
+impl fmt::Display for MaclItem {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if !self.grantee.is_public() {
+            write!(f, "{}", self.grantee)?;
+        }
+        write!(f, "={}/{}", self.acl_mode, self.grantor)
+    }
+}
+
```

Comment:
It would be nice to implement `ToSql` and `FromSql`, but I don't think it's possible because we don't have access to the type in `pgrepr`.